libyul/AsmPrinter.cpp

```diff
@@ -41,48 +41,60 @@ using namespace solidity;
 using namespace solidity::util;
 using namespace solidity::yul;
 
-//@TODO source locations
-
-string AsmPrinter::operator()(Literal const& _literal) const
+string AsmPrinter::operator()(Literal const& _literal)
 {
+	string const locationComment = formatSourceLocationComment(_literal.debugData, !m_insideExpression);
+
 	switch (_literal.kind)
 	{
 	case LiteralKind::Number:
 		yulAssert(isValidDecimal(_literal.value.str()) || isValidHex(_literal.value.str()), "Invalid number literal");
-		return _literal.value.str() + appendTypeName(_literal.type);
+		return locationComment + _literal.value.str() + appendTypeName(_literal.type);
 	case LiteralKind::Boolean:
 		yulAssert(_literal.value == "true"_yulstring || _literal.value == "false"_yulstring, "Invalid bool literal.");
-		return ((_literal.value == "true"_yulstring) ? "true" : "false") + appendTypeName(_literal.type, true);
+		return locationComment + ((_literal.value == "true"_yulstring) ? "true" : "false") + appendTypeName(_literal.type, true);
 	case LiteralKind::String:
 		break;
 	}
 
-	return escapeAndQuoteString(_literal.value.str()) + appendTypeName(_literal.type);
+	return locationComment + escapeAndQuoteString(_literal.value.str()) + appendTypeName(_literal.type);
 }
 
-string AsmPrinter::operator()(Identifier const& _identifier) const
+string AsmPrinter::operator()(Identifier const& _identifier)
 {
 	yulAssert(!_identifier.name.empty(), "Invalid identifier.");
-	return _identifier.name.str();
+	return formatSourceLocationComment(_identifier.debugData, !m_insideExpression) + _identifier.name.str();
 }
 
-string AsmPrinter::operator()(ExpressionStatement const& _statement) const
+string AsmPrinter::operator()(ExpressionStatement const& _statement)
 {
-	return std::visit(*this, _statement.expression);
+	string const locationComment = formatSourceLocationComment(_statement.debugData, true);
+	m_insideExpression++;
```

Comment:
You don't need this member. You can tell from the ast node kind whether or not you are inside an expression.

FunctionCall, Identifier and Literal are inside expressions, the rest are statements - see the definition of Expression in ASTForward.h

Reply:
FunctionCall can also be outside an expression, though. And technically, an identifier/literal could also be on its own?